blockchain/blockchain.go

```diff
@@ -611,6 +611,11 @@ func (bc *BlockChain) setHeadBeyondRoot(head uint64, root common.Hash, repair bo
 			// to low, so it's safe the update in-memory markers directly.
 			bc.currentFastBlock.Store(newHeadFastBlock)
 		}
+
+		// Rewind the supply checkpoint
+		newLastSupplyCheckpointNumber := header.Number.Uint64() - (header.Number.Uint64() % params.SupplyCheckpointInterval)
+		bc.db.WriteLastSupplyCheckpointNumber(newLastSupplyCheckpointNumber)
```

Comment:
1. This unconditionally stores the nearest check point number. Is it still safe if the rewinding number is beyond the nearest check pointer number because check point key may be deleted more than this range in `delFn`.

2. It works with a case below
------------ c1(exist) ------------ c2(exist) ------------ c3(exist) ------------ headerNumber ------------

is it safe?
------------ c1(exist) ------------ c2(exist) ------------ c3(non-exist) ------------ headerNumber ------------

* c = checkpoint

maybe we do not need to care much about this because `accumulateReward` will be triggered?

Reply:
a. About the feasibility of the situation,
  Suppose c1=128, c2=256, c3=384.
  How can we arrive that situation? start with currentHeader=400, SetHead(270), then immediately SetHead(380)?

b. For recovery,
  - `catchup()` is always (re)started after sethead
    - When setHead is performed at node startup (via --start-block-num or hardfork config mismatch, etc) then LastNum is rewinded by `updateFn` then `catchup()` starts.
    - When setHead is triggered by debug_setHead API, then APIBackend.SetHead Stop() and Start() the SupplyManager.
  - When `catchup()` starts, it checks that the SupplyCheckpoint(num = LastCheckpointNumber()) exists and if not, the goroutine will restart from the genesis.

Reply:
a. Given the situation, I thought that last supply checkpoint number would be 384, but some of keys were deleted more than 14. However, the `updateFn` is called for every block number (I confused the delete function was only invoked for every block number without looking the detail). So, it was false alarm. Sorry for noise.

b.  Starting from genesis may be heavy work if it does not occur rarely. For example, we did `debug_setHead(klay.blockNumber - 1)`, which involves to modify the last checkpoint to `lastcheckpoint -  (headerNumber % lastchecjpoint)` and if there's no corresponding key with this number, then reading last checkpoint will be zero, result in building from scratch, but I 'm not sure where this case does exist (e.g., c1, c2 exists, c3 non-exist, c4 exist).

Reply:
a. Yes that's right.
b. If you see the upper part of `updateFn`, if there state missing the header it rewinds to genesis because you encountered an unexpected "gap". I'd employ the same strategy.